cmd/dockerd/daemon.go

```diff
@@ -251,6 +261,30 @@ func (cli *DaemonCli) start(opts daemonOptions) (err error) {
 		logrus.Fatalf("Error starting cluster component: %v", err)
 	}
 
+	builderStateDir := filepath.Join(cli.Config.Root, "builder")
+
+	fsCache, err := fscache.NewFSCache(fscache.Opt{
+		Backend: fscache.NewNaiveCacheBackend(builderStateDir),
+		Root:    builderStateDir,
+		GCPolicy: fscache.GCPolicy{ // TODO: expose this in config
+			MaxSize:         1024 * 1024 * 512,  // 512MB
+			MaxKeepDuration: 7 * 24 * time.Hour, // 1 week
+		},
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to create fscache")
+	}
+
+	manager, err := dockerfile.NewBuildManager(d, sm, fsCache)
+	if err != nil {
+		return err
+	}
+
+	bb, err := buildbackend.NewBackend(d, manager, fsCache)
+	if err != nil {
+		return errors.Wrap(err, "failed to create buildmanager")
+	}
```

Comment:
`initRouter()` now takes 6 args, which is way too many.

An options struct for `initRouter()` would be good. Then all of this code could be moved to a function that creates the option struct.